Add a PEG (Parsing Expression Grammar) lexer

goodmami opened the pull request:
[Parsing Expression Grammars](https://en.wikipedia.org/wiki/Parsing_expression_grammar) (PEGs) have been around for about 15 years and they are fairly popular these days for describing non-ambiguous grammars, such as for programming languages and data formats.

This lexer (`pygments.lexers.grammar_notation.PegLexer`) builds on the original PEG syntax described by Bryan Ford (see [here](https://bford.info/pub/lang/peg.pdf)) with some common extensions seen in various implementations of PEG, such as (optionally) using `|` for choices instead of `/`, `=` or `:` instead of `<-` for rule definitions, cut operators (`^` or `~`), and string modifiers, such as `r"a regex"`.

I've added `tests/test_grammar_notation.py`, but only included tests relevant to this PR instead of going back and adding tests for `BnfLexer`, etc.

Anteru commented on pygments/lexers/grammar_notation.py:
```diff
@@ -211,3 +211,58 @@ class JsgfLexer(RegexLexer):
+    * A single `a-z` character immediately before a string, or
+      multiple `a-z` characters following a string, are part of the
+      string (e.g., `r"..."` or `"..."ilmsuxa`).
+    """
```
Can you please add ``.. versionadded:: 2.6`` at the end?

goodmami commented:
Added the `versionadded` tag and added the language to `doc/languages.rst`, as requested.

There are two remaining things about the lexer that I thought about and I'll write them here for posterity:
* At least one syntax variant (used in [Parsimonious](https://github.com/erikrose/parsimonious)) uses `~"..."`, `~r"..."`, etc. for regular expressions, as an extension. This was not compatible with `~` as a cut operator (which is used in the new [pegen](https://github.com/gvanrossum/pegen) and older, but not exactly PEG, [TatSu](https://github.com/neogeny/TatSu) libraries), but the impact is minimal: it will just be highlighted differently and does not break parsing.
* I did not account for extensions that put semantic actions between `{` and `}` at the end of the rule (as in the previously mentioned pegen, as well as [Rats!](https://cs.nyu.edu/rgrimm/xtc/rats-intro.html) and some other parsers. I chose not to do this because I didn't see such support in the other grammar notation lexers, but I can add it if requested.

goodmami commented:
@Anteru do you wish me to resolve the conflicts in `pygments/lexers/_mapping.py` or does the person merging handle this? And is there anything else you're waiting on me for?

Anteru commented:
Hi, that would be appreciated, but I regenerate the `_mapping.py` file quite often myself already. I just didn't get around to merging this yet, will try to do it this week.

goodmami commented:
Ok I've rebased to the current master branch. Let me know if you need anything else.

Anteru commented:
Merged, thanks for the contribution!
